### PR TITLE
#274 Admin Global Search: Search categories

### LINKED ISSUE #274: Admin Global Search: Search categories
Hello :helicopter: 

Global search in backend can search into products, orders, customers, pages and config sections.
But categories cannot be reached.

I think the difficulty here depends on creating URLs for category admin pages.

### PR DESCRIPTION
**Pull request by FredericMartinez**
### Description (*)
Global search in backend can search into products, orders, customers, pages and config sections.
Now we can search categories too.

### Fixed Issues (if relevant)
1. Fixes https://github.com/magento/community-features/issues/274

### Manual testing scenarios (*)
1. Login into backend
2. Create a test category "Try to find me"
3. Use global search (in header to right, next to admin account setting)
4. Enter "to find"
5. You will see in autocomplete list our new fresh category, link is valid and redirect to the edit form

### Contribution checklist (*)
 - [X] Pull request has a meaningful description of its purpose
 - [X] All commits are accompanied by meaningful commit messages
 - [X] All new or changed code is covered with unit/integration tests (if applicable)
 - [X] All automated tests passed successfully (all builds are green)

### CODE REVIEW

**Comment by FredericMartinez** [comment]
@magento give me test instance

**Comment by magento-engcom-team** [comment]
Hi @FredericMartinez. Thank you for your request. I'm working on Magento instance for you

**Comment by magento-engcom-team** [comment]
Hi @FredericMartinez, here is your [new Magento instance](https://pr-28588.instances.magento-community.engineering).
Admin access: [https://pr-28588.instances.magento-community.engineering/admin_d843](https://pr-28588.instances.magento-community.engineering/admin_d843)
Login: `6e383ff3` Password: `6b137e49e023`
Instance will be terminated in up to 3 hours.

**Comment by FredericMartinez** [comment]
It works like expected

**Comment by okorshenko** [comment]
@magento run Unit Tests

**Comment by Stepa4man** [comment]
@magento give me test instance

**Comment by magento-engcom-team** [comment]
Hi @Stepa4man. Thank you for your request. I'm working on Magento instance for you

**Comment by Stepa4man** [comment]
Hey, @okorshenko 
What do you think if we ask the product owner about it?
It's definitely a good improvement. I am really happy about it.
But... I'm not sure about UI of this solution. 

My concern here about the following:
Global search doesn't search at the admin panel doesn't really search anything, it just redirects admin user to grid page of particular entity type, and submits a search with the requested string there.
But for Category, we don't have a grid and search there. So @FredericMartinez implemented search action directly on the global search level. Which does its job, but just doesn't match the basic concept of this tool.

**Comment by Stepa4man** [comment]
Hi @FredericMartinez 
Thanks for your contribution.

I want to ask you to cover implemented functionality by tests according to [DOD](https://devdocs.magento.com/guides/v2.3/contributor-guide/contributing_dod.html)

Also please  add to all changed (created) php files
```php
    declare(strict_types=1);
```
You can read more [here](https://devdocs.magento.com/guides/v2.3/coding-standards/technical-guidelines.html#1-basic-programming-principles)

File: app/code/Magento/CatalogSearch/Model/Search/Category.php [hunk 1]
```diff
@@ -0,0 +1,121 @@
...
+        \Magento\Backend\Helper\Data $adminhtmlData,
+        \Magento\Catalog\Api\CategoryListInterface $categoryRepository,
+        \Magento\Framework\Api\SearchCriteriaBuilder $searchCriteriaBuilder,
+        \Magento\Framework\Api\FilterBuilder $filterBuilder,
+        \Magento\Framework\Stdlib\StringUtils $string
+    )
+    {
+        $this->_adminhtmlData = $adminhtmlData;
+        $this->categoryRepository = $categoryRepository;
+        $this->searchCriteriaBuilder = $searchCriteriaBuilder;
```
Comment (Stepa4man) [inline app/code/Magento/CatalogSearch/Model/Search/Category.php:71]: Please use searchCriteriaBuilderFactory here

File: app/code/Magento/CatalogSearch/Model/Search/Category.php [hunk 2]
```diff
@@ -0,0 +1,121 @@
...
+        $this->searchCriteriaBuilder->setCurrentPage($this->getStart());
+        $this->searchCriteriaBuilder->setPageSize($this->getLimit());
+        $searchFields = ['name'];
+
+        $filters = [];
+        foreach ($searchFields as $field) {
+            $filters[] = $this->filterBuilder
+                ->setField($field)
+                ->setConditionType('like')
+                ->setValue('%' . $this->getQuery() . '%')
```
Comment (Stepa4man) [inline app/code/Magento/CatalogSearch/Model/Search/Category.php]: Please use `sprintf` here

File: app/code/Magento/CatalogSearch/Model/Search/Category.php [hunk 3]
```diff
@@ -0,0 +1,121 @@
...
+        }
+        $this->searchCriteriaBuilder->addFilters($filters);
+
+        $searchCriteria = $this->searchCriteriaBuilder->create();
+        $searchResults = $this->categoryRepository->getList($searchCriteria);
+
+        foreach ($searchResults->getItems() as $category) {
+            $description = strip_tags($category->getDescription());
+            $result[] = [
+                'id' => 'category/1/' . $category->getId(),
```
Comment (Stepa4man) [inline app/code/Magento/CatalogSearch/Model/Search/Category.php]: Please use sprintf here

File: app/code/Magento/CatalogSearch/Model/Search/Category.php [hunk 4]
```diff
@@ -0,0 +1,121 @@
...
+     * @var \Magento\Framework\Api\FilterBuilder
+     */
+    protected $filterBuilder;
+
+    /**
+     * Magento string lib
+     *
+     * @var \Magento\Framework\Stdlib\StringUtils
+     */
+    protected $string;
```
Comment (Stepa4man) [inline app/code/Magento/CatalogSearch/Model/Search/Category.php]: Please use a private access modifier

File: app/code/Magento/CatalogSearch/Model/Search/Category.php [hunk 5]
```diff
@@ -0,0 +1,121 @@
...
+ * @api
+ */
+class Category extends \Magento\Framework\DataObject
+{
+    /**
+     * Adminhtml data
+     *
+     * @var \Magento\Backend\Helper\Data
+     */
+    protected $_adminhtmlData = null;
```
Comment (Stepa4man) [inline app/code/Magento/CatalogSearch/Model/Search/Category.php]: Please don't user underscore in property name it's an outdated approach

File: app/code/Magento/CatalogSearch/Model/Search/Category.php [hunk 6]
```diff
@@ -0,0 +1,121 @@
...
+     * @param \Magento\Framework\Api\SearchCriteriaBuilder $searchCriteriaBuilder
+     * @param \Magento\Framework\Api\FilterBuilder $filterBuilder
+     * @param \Magento\Framework\Stdlib\StringUtils $string
+     */
+    public function __construct(
+        \Magento\Backend\Helper\Data $adminhtmlData,
+        \Magento\Catalog\Api\CategoryListInterface $categoryRepository,
+        \Magento\Framework\Api\SearchCriteriaBuilder $searchCriteriaBuilder,
+        \Magento\Framework\Api\FilterBuilder $filterBuilder,
+        \Magento\Framework\Stdlib\StringUtils $string
```
Comment (Stepa4man) [inline app/code/Magento/CatalogSearch/Model/Search/Category.php]: Please import all classes

File: app/code/Magento/CatalogSearch/Model/Search/Category.php [hunk 7]
```diff
@@ -0,0 +1,121 @@
...
+
+    /**
+     * Magento string lib
+     *
+     * @var \Magento\Framework\Stdlib\StringUtils
+     */
+    protected $string;
+
+    /**
+     * Initialize dependencies.
```
Comment (Stepa4man) [inline app/code/Magento/CatalogSearch/Model/Search/Category.php]: Not sure if we need comment here.

File: app/code/Magento/CatalogSearch/Model/Search/Category.php [hunk 8]
```diff
@@ -0,0 +1,121 @@
...
+ * @method int|null getStart()
+ * @method bool hasStart()
+ * @method Category setLimit(int $limit)
+ * @method int|null getLimit()
+ * @method bool hasLimit()
+ * @method Category setResults(array $results)
+ * @method array getResults()
+ * @api
+ */
+class Category extends \Magento\Framework\DataObject
```
Comment (Stepa4man) [inline app/code/Magento/CatalogSearch/Model/Search/Category.php]: Please import parent class

**Review by Stepa4man — CHANGES_REQUESTED** [review]
@FredericMartinez 
Please check out my comments.
I know you just implemented it as \Magento\CatalogSearch\Model\Search\Catalog
But it's quite an old class, If you have doubts about how to write code you can read [this documentation](https://devdocs.magento.com/guides/v2.3/coding-standards/technical-guidelines.html)

**Comment by okorshenko** [comment]
Hi @Stepa4man 
That's interesting PR. Let's review it with PO on Monday. I will send you an invitation for next triage session

**Comment by FredericMartinez** [comment]
@Stepa4man Thanks you for your review.

I understand all your comments about style codings
But, my Magento example are: `Magento\Backend\Model\Search\Customer`

You can find almost a copy/paste from this class with a category filter instead of a customer filter.

Actually, I hesitated to write the class properly, but I assume to have the same old code style.
Indeed, when you will rewrite Magento\Backend\Model\Search\Customer and all the other Admin Global Search classes, you will be able to clean them all at one time, my PR included.

My point was to contributed as it, it works like the old stylish classes and I can get my git patch from this PR in my projects.

Thanks for your time @Stepa4man & @okobchenko 
Have a nice week.

**Comment by Stepa4man** [comment]
Hi @FredericMartinez, I understand your point but you should also understand us.
Yes, we definitely have some legacy code that requires cleaning up. But this is not the reason to deliver a new bad quality code. We have [Magento coding standards](https://devdocs.magento.com/guides/v2.3/coding-standards/technical-guidelines.html)  and [DoD](https://devdocs.magento.com/guides/v2.3/contributor-guide/contributing_dod.html) and PR can be merged only if it follows these documents.
Thanks.

**Comment by engcom-Kilo** [comment]
Working on this .

**Comment by engcom-Kilo** [comment]
Added correction for code.

**Comment by engcom-Kilo** [comment]
Added MFTF test.

**Comment by engcom-Kilo** [comment]
@magento run all tests

**Comment by engcom-Hotel** [comment]
Hello @Stepa4man could you please review PR changes.
Thank you.

**Review by sidolov — APPROVED** [review]

**Comment by magento-engcom-team** [comment]
Hi @sidolov, thank you for the review.
[ENGCOM-8204](https://jira.corp.magento.com/browse/ENGCOM-8204) has been created to process this Pull Request

**Comment by engcom-Hotel** [comment]
@magento run all tests

**Comment by engcom-Alfa** [comment]
:heavy_check_mark: QA Passed

**Preconditions:**

Magento 2.4-develop
Sample Data 

**Manual testing scenario:**

1. Login into backend;
2. Use global search (in the header to right, next to admin account setting)
3. Enter "Luma Yoga"

**Before:**  :heavy_multiplication_x: We don't see the specified category in the search list



**After:** :heavy_check_mark: we see in the autocomplete list our new fresh category, a link is valid and redirects to the edit form



Peek 2020-09-22 12-33

**Comment by engcom-Hotel** [comment]
@magento run all tests

**Comment by engcom-Hotel** [comment]
@magento run all tests

**Pull request merged**